File: mockserver/mockserver-core/src/main/java/org/mockserver/load/LoadProfile.java
package org.mockserver.load;

import org.mockserver.model.ObjectWithJsonToString;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * The load profile of a {@link LoadScenario}: an ordered list of {@link LoadStage}s the orchestrator
 * runs <em>in sequence</em>. Each stage holds or ramps a setpoint — concurrent virtual users
 * (closed model, {@link LoadStageType#VU}), an arrival rate in iterations/second (open model,
 * {@link LoadStageType#RATE}), or no load at all ({@link LoadStageType#PAUSE}) — for its
 * {@code durationMillis}, optionally shaped by a {@link RampCurve}.
 *
 * <p>The total run duration is the sum of the stage durations; the orchestrator advances stage by
 * stage and ends after the last one (or when {@code maxRequests} is hit, or on stop). Per-stage
 * setpoint functions ({@link LoadStage#targetVusAt(long)} / {@link LoadStage#targetRateAt(long)}) are
 * pure and deterministic so ramp progression is unit-testable without driving traffic.
 *
 * <p>A profile may instead carry a declarative {@link LoadShape} (a named SPIKE / STAIRS / RAMP_HOLD
 * pattern). When a {@code shape} is set and no explicit {@code stages} are present,
 * {@link #getStages()} returns the stages {@link LoadShapes#expand expanded} from the shape — so the
 * orchestrator, which only ever calls {@link #getStages()}, needs no change. Explicit stages always
 * win: if both are set the shape is ignored.
 */
public class LoadProfile extends ObjectWithJsonToString {

    private List<LoadStage> stages = new ArrayList<>();
    private LoadShape shape;
    /** Lazily-computed expansion of {@link #shape}; never serialized (derived, not stored). */
    private transient volatile List<LoadStage> expandedShapeStages;

    public LoadProfile() {
    }

    public static LoadProfile loadProfile() {
        return new LoadProfile();
    }

    public static LoadProfile of(LoadStage... stages) {
        LoadProfile profile = new LoadProfile();
        Collections.addAll(profile.stages, stages);
        return profile;
    }

    /** Convenience: a single constant-VU stage. */
    public static LoadProfile constant(int vus, long durationMillis) {
        return of(LoadStage.constantVus(vus, durationMillis));
    }

    /** Convenience: a single linear VU ramp stage. */
    public static LoadProfile linear(int startVus, int endVus, long durationMillis) {
        return of(LoadStage.rampVus(startVus, endVus, durationMillis, RampCurve.LINEAR));
    }

    /** Convenience: a single constant arrival-rate (iterations/second) stage. */
    public static LoadProfile constantRate(double rate, long durationMillis) {
        return of(LoadStage.constantRate(rate, durationMillis));
    }

    /** Convenience: a named load shape. */
    public static LoadProfile shaped(LoadShape shape) {
        return new LoadProfile().withShape(shape);
    }

    /**
     * The ordered stages the orchestrator runs. Explicit {@link #stages} always win; otherwise, when a
     * {@link #shape} is set, this returns its {@link LoadShapes#expand expansion} (computed once and
     * cached). With neither, this is the (empty) explicit stages list.
     */
    public List<LoadStage> getStages() {
        if (stages != null && !stages.isEmpty()) {
            return stages;
        }
        if (shape != null) {
            if (expandedShapeStages == null) {
                // Unsynchronised check-then-set is intentionally benign here: expandedShapeStages is
                // volatile and LoadShapes.expand is a pure, deterministic, idempotent function of the
                // (immutable-for-the-lifetime-of-this-shape) shape, so a rare concurrent double-compute
                // only ever yields an equivalent value, and the volatile write safely publishes it.
                // This is off the hot path, so a lock would add cost without buying correctness.
                expandedShapeStages = LoadShapes.expand(shape);
            }
            return expandedShapeStages;
        }
        return stages;
    }

    /** The explicit stages exactly as set (never the shape expansion) — used by serialization. */
    public List<LoadStage> getRawStages() {
        return stages;
    }

    public LoadShape getShape() {
        return shape;
    }

    public LoadProfile withShape(LoadShape shape) {
        this.shape = shape;
        this.expandedShapeStages = null;
        return this;
    }

    public LoadProfile withStages(List<LoadStage> stages) {
        this.stages = stages != null ? stages : new ArrayList<>();
        return this;
    }

    public LoadProfile withStages(LoadStage... stages) {
        this.stages = new ArrayList<>();
        Collections.addAll(this.stages, stages);
        return this;
    }

    public LoadProfile addStage(LoadStage stage) {
        if (this.stages == null) {
            this.stages = new ArrayList<>();
        }
        this.stages.add(stage);
        return this;
    }

    /** Sum of all stage durations — the total run length (reflects the shape expansion). */
    public long totalDurationMillis() {
        long total = 0;
        for (LoadStage stage : getStages()) {
            total += Math.max(0, stage.getDurationMillis());
        }
        return total;
    }

    /** The maximum VU count any stage requests, used to enforce the VU hard cap up-front. */
    public int peakVus() {
        int peak = 0;
        for (LoadStage stage : getStages()) {
            peak = Math.max(peak, stage.peakVus());
        }
        return peak;
    }

    /** The maximum arrival rate any stage requests, used to enforce the rate hard cap up-front. */
    public double peakRate() {
        double peak = 0.0;
        for (LoadStage stage : getStages()) {
            peak = Math.max(peak, stage.peakRate());
        }
        return peak;
    }
}

File: mockserver-client-php/src/LoadProfile.php
<?php

declare(strict_types=1);

namespace MockServer;

/**
 * The load profile of a {@see LoadScenario}: an ordered list of
 * {@see LoadStage}s the orchestrator runs in sequence. Each stage holds or ramps
 * a setpoint — concurrent virtual users (VU, closed model), an arrival rate in
 * iterations/second (RATE, open model), or no load at all (PAUSE) — for its
 * duration.
 *
 * Use {@see LoadProfile::of()} to build from a list of stages, or the
 * convenience constructors {@see LoadProfile::constant()} /
 * {@see LoadProfile::linear()} for a single VU stage.
 *
 * @example
 *   LoadProfile::of(
 *       LoadStage::vuRamp(1, 10, 10000),
 *       LoadStage::vuHold(10, 30000),
 *       LoadStage::pause(5000),
 *   );
 */
class LoadProfile implements \JsonSerializable
{
    /** @var array<int, LoadStage> */
    private array $stages = [];
    private ?LoadShape $shape = null;

    public function __construct()
    {
    }

    /**
     * Build a profile from an ordered list of stages.
     */
    public static function of(LoadStage ...$stages): self
    {
        $profile = new self();
        $profile->stages = $stages;

        return $profile;
    }

    /**
     * Build a profile from a single declarative named {@see LoadShape} that the
     * server expands into ordinary stages. Use a shape OR an explicit list of
     * stages, not both.
     */
    public static function fromShape(LoadShape $shape): self
    {
        $profile = new self();
        $profile->shape = $shape;

        return $profile;
    }

    /**
     * A single VU stage holding {@code $vus} virtual users for the whole
     * {@code $durationMillis}.
     */
    public static function constant(int $vus, int $durationMillis): self
    {
        return self::of(LoadStage::vuHold($vus, $durationMillis));
    }

    /**
     * A single linear VU ramp from {@code $startVus} to {@code $endVus} over
     * {@code $durationMillis}.
     */
    public static function linear(int $startVus, int $endVus, int $durationMillis): self
    {
        return self::of(LoadStage::vuRamp($startVus, $endVus, $durationMillis, 'LINEAR'));
    }

    /**
     * Append a stage and return this profile.
     */
    public function addStage(LoadStage $stage): self
    {
        $this->stages[] = $stage;

        return $this;
    }

    /**
     * Set the declarative named {@see LoadShape} for this profile. If explicit
     * stages are also set the server prefers the stages.
     */
    public function shape(LoadShape $shape): self
    {
        $this->shape = $shape;

        return $this;
    }

    /**
     * @return array<string, mixed>
     */
    public function jsonSerialize(): array
    {
        return $this->toArray();
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [];
        if ($this->stages !== []) {
            $result['stages'] = array_map(
                static fn (LoadStage $stage): array => $stage->toArray(),
                $this->stages,
            );
        }
        if ($this->shape !== null) {
            $result['shape'] = $this->shape->toArray();
        }
        // Preserve the historical shape: a profile built from stages always
        // emits a (possibly empty) "stages" array even when none were added.
        if ($this->stages === [] && $this->shape === null) {
            $result['stages'] = [];
        }

        return $result;
    }
}
